Allow editing of Repositories using the form-yaml experience

debsmita1 opened the pull request:
**JIRA story:**
https://issues.redhat.com/browse/ODC-6726

**Solution description:**
using the existing form for adding helm chart repositories to edit them as well

**GIF:**
https://user-images.githubusercontent.com/22490998/183068664-ff457ff2-f42a-4561-9b26-95fdce01e52f.mov

**Test coverage:**
<img width="573" alt="Screenshot 2022-07-22 at 5 45 06 PM" src="https://user-images.githubusercontent.com/22490998/180440874-1ea8b444-a22c-4428-9ce4-cd22ed31778f.png">

debsmita1 commented:
/cc @lokanandaprabhu 
/assign @jerolimov

divyanshiGupta commented:
@debsmita1 please rebase the PR. I have reviewed the code and it looks good to me but need to test this once before adding lgtm.

divyanshiGupta commented:
@debsmita1 the `Edit helm chart repository` action on the repository detail page opens the yaml tab instead of the edit form. Is this expected?

debsmita1 commented:
> @debsmita1 the `Edit helm chart repository` action on the repository detail page opens the yaml tab instead of the edit form. Is this expected?

Thanks for catching this. Fixed it ! PTAL

jerolimov commented on frontend/packages/helm-plugin/src/components/forms/HelmChartRepository/CreateHelmChartRepositoryForm.tsx:
```diff
@@ -51,17 +55,13 @@ const CreateHelmChartRepositoryForm: React.FC<FormikProps<FormikValues> & {
           : HelmChartRepositoryModel
       }
-      showSamples={!resource}
+      showSamples
```
For other resources we show the samples only when creating a new resource, not in the edit flow.

Can we archive this by checking `existingRepo` here? Btw: Why `existingRepo` instead of `resource`, isn't it the same?

```suggestion
      showSamples={!existingRepo}
```

jerolimov commented on frontend/packages/helm-plugin/console-extensions.json:
```diff
@@ -132,7 +160,7 @@
       "id": "project-helm-chart-repositories",
       "groupId": "helm-chart-repositories",
-      "href": "/ns/:namespace/helmchartrepositories/~new",
+      "href": "/ns/:namespace/helmchartrepositories/~new?&actionOrigin=addPage",
```
Nit:

```suggestion
      "href": "/ns/:namespace/helmchartrepositories/~new?actionOrigin=addPage",
```

jerolimov commented on frontend/packages/helm-plugin/src/catalog/components/HelmCatalogTypeDescription.tsx:
```diff
@@ -19,7 +19,7 @@ const LinkToCreatePHCR: React.FC = () => {
       <Trans ns="helm-plugin">
         Alternatively, developers can{' '}
-        <Link to={`/ns/${namespace}/helmchartrepositories/~new`}>
+        <Link to={`/ns/${namespace}/helmchartrepositories/~new?&actionOrigin=catalog`}>
```
Nit:

```suggestion
        <Link to={`/ns/${namespace}/helmchartrepositories/~new?actionOrigin=catalog`}>
```

jerolimov commented on frontend/packages/helm-plugin/src/actions/providers.ts:
```diff
@@ -65,3 +74,19 @@ export const useTopologyActionProvider = ({
+  const [kindObj, inFlight] = useK8sModel(referenceFor(resource));
+  const { t } = useTranslation();
+  const actions = React.useMemo(() => {
+    let commonActions = getCommonResourceActions(kindObj, resource);
+    const index = commonActions.findIndex((action: Action) => action.id === 'edit-resource');
+    if (index) {
+      commonActions = commonActions.filter((action: Action) => action.id !== 'edit-resource');
+      commonActions.splice(index, 0, editChartRepository(kindObj, resource, t));
+    }
+    return commonActions;
```
Do you want check here if the index exist?

Array.findIndex returns -1 if it doesn't exist and would run your if statement.

And it returns 0 if the first action is edit-resource and this would jump over your if statement.

You should also be very careful with Array.splice as it doesn't return a new array. That's only okay here because of the filter one line above. I always recommend to use Array.slice instead also it's not required here.

```suggestion
    let commonActions = getCommonResourceActions(kindObj, resource);
    const index = commonActions.findIndex((action: Action) => action.id === 'edit-resource');
    if (index => 0) {
      commonActions = commonActions.filter((action: Action) => action.id !== 'edit-resource');
      commonActions = commonActions.slice(index, 0, editChartRepository(kindObj, resource, t));
    }
    return commonActions;
```

jerolimov commented on frontend/packages/helm-plugin/src/components/forms/HelmChartRepository/CreateHelmChartRepositoryForm.tsx:
```diff
@@ -72,13 +72,21 @@ const CreateHelmChartRepositoryForm: React.FC<FormikProps<FormikValues> & {
     !showScopeType && formData.scope === ProjectHelmChartRepositoryModel.kind
-      ? t('helm-plugin~Add helm chart repository in the namespace.')
-      : t('helm-plugin~Add helm chart repository');
+      ? t('helm-plugin~{{formDescType}} helm chart repository in the namespace.', {
+          formDescType: existingRepo ? 'Update' : 'Add',
+        })
+      : t('helm-plugin~{{formDescType}} helm chart repository.', {
+          formDescType: existingRepo ? 'Update' : 'Add',
+        });
```
I understand the benefit for a developer of concatenating strings like `Create`/`Edit` plus `ProjectHelmChartRepository`/`Helm Chart Repository` - or `Add`/`Update` plus a sentence.

The problem here is that this expects that all translations have the form `$verb $noun` or `$verb sentence`, but that's not required in all languages and I would expect that makes the life of the translation team more difficult.

Using variables here also will not allow the translation team to translate "Update" and "Add".

So I recommend a much more simple (you can also call it stupid) code like this:

```suggestion
  const formDescription =
    !showScopeType && formData.scope === ProjectHelmChartRepositoryModel.kind
      ? (existingRepo
          ? t('helm-plugin~Update helm chart repository in the namespace.')
          : t('helm-plugin~Add helm chart repository in the namespace.')
        )
      : (existingRepo
          ? t('helm-plugin~Update helm chart repository.')
          : t('helm-plugin~Add helm chart repository.')
        );
```

jerolimov commented on frontend/packages/helm-plugin/src/components/forms/HelmChartRepository/CreateHelmChartRepositoryForm.tsx:
```diff
@@ -72,13 +72,21 @@ const CreateHelmChartRepositoryForm: React.FC<FormikProps<FormikValues> & {
     !showScopeType && formData.scope === ProjectHelmChartRepositoryModel.kind
-      ? t('helm-plugin~Create ProjectHelmChartRepository')
-      : t('helm-plugin~Create Helm Chart Repository');
+      ? t('helm-plugin~{{formType}} ProjectHelmChartRepository', {
+          formType: existingRepo ? 'Edit' : 'Create',
+        })
+      : t('helm-plugin~{{formType}} Helm Chart Repository', {
+          formType: existingRepo ? 'Edit' : 'Create',
+        });
```
Same translation issue here. Edit and Create should not be variables. This should be part of the translation. Keep it simple:

```suggestion
      ? (
        existingRepo
          ? t('helm-plugin~Edit ProjectHelmChartRepository')
          : t('helm-plugin~Create ProjectHelmChartRepository')
      )
      : (
        existingRepo
          ? t('helm-plugin~Edit Helm Chart Repository')
          : t('helm-plugin~Create Helm Chart Repository')
      );
```

jerolimov commented on frontend/packages/helm-plugin/src/components/list-page/HelmChartRepositoryRow.tsx:
```diff
@@ -0,0 +1,46 @@
+      </TableData>
+      <TableData>{obj.spec?.name ?? '-'}</TableData>
+      <TableData>{t('helm-plugin~All namespaces')}</TableData>
+      <TableData>{obj.spec?.disabled ?? '-'}</TableData>
```
The other places uses a translation here:

```suggestion
      <TableData>{obj.spec?.disabled ? t('helm-plugin~True') : t('helm-plugin~False')}</TableData>
```

jerolimov reviewed (changes requested):
I reviewed the 2nd commit and found some nits and some translation issues. Please see the comments above.

The direction of this change is fine, so I would like to approve this right now. Please check with the final reviewer that all comments are addressed or let me know if I should take a new look at this.

/approve

jerolimov commented:
Found also another UI issue: The LazyActionMenu is not shown correctly above the Add/Remove from navigation button.

Because this is a general UI issue and an edge case I opened a [BZ ticket](https://bugzilla.redhat.com/show_bug.cgi?id=2115265) for this.

https://user-images.githubusercontent.com/139310/182818276-e618bbf8-9541-40ee-97bb-d3cbe881647a.mp4

https://user-images.githubusercontent.com/139310/182818264-62b069a9-2e84-46df-a923-b160cdf8d6c6.mp4

invincibleJai commented:
/hold

adding hold as this PR is dependent on https://github.com/openshift/console/pull/11844, feel free to unhold it once the dependent PR is in.

cc @debsmita1

jerolimov commented:
#11844 is merged

/unhold

jerolimov reviewed (changes requested):
As posted on slack, the helm chart empty state page contains a link to the catalog but links to the namespace "undefined".

vikram-raj commented:
For both the Edit HCR and PHCR. Form heading is the same `Edit Helm Chart Repository`. Is it expected?

![Uploading Kapture 2022-08-05 at 14.59.52.gif…]()

vikram-raj commented:
In the case of Edit HCR, the Form page URL have `undefined` as a namespace name. IMO as HCR is not a namespace scoped so can remove the `ns/:namespace` from the URL and add `/k8s/cluster/` instead. wdyt?

<img width="1086" alt="image" src="https://user-images.githubusercontent.com/2561818/183051987-b1c81eaf-cf4c-4557-97b1-e61acd0fa87c.png">

debsmita1 commented:
> In the case of Edit HCR, the Form page URL have `undefined` as a namespace name. IMO as HCR is not a namespace scoped so can remove the `ns/:namespace` from the URL and add `/k8s/cluster/` instead. wdyt?
> 
> <img alt="image" width="1086" src="https://user-images.githubusercontent.com/2561818/183051987-b1c81eaf-cf4c-4557-97b1-e61acd0fa87c.png">

Fixed it ! and updated the video in the PR description. PTAL

vikram-raj reviewed (approved):
Verified the changes and it works as expected.

sanketpathak commented on frontend/packages/helm-plugin/integration-tests/features/helm/helm-page-tabs.feature:
```diff
@@ -1,4 +1,4 @@
-@helm @ODC6715
+@helm @ODC6685
```
```suggestion
@helm @odc-6685
```

sanketpathak commented on frontend/packages/helm-plugin/integration-tests/features/helm/helm-page-tabs.feature:
```diff
@@ -38,4 +38,54 @@ Feature: Add repositories tab in Helm navigation item
+               And user edits "helm-test1" "ProjectHelmChartRepository"
+               And user is at "Edit ProjectHelmChartRepository" form
+               And user enters Display name as "My charts"
+               And user clicks on Save button
+             Then user can see "ProjectHelmChartRepository" "helm-test1" details page
```
```suggestion
               And user clicks on Save button
              Then user can see "ProjectHelmChartRepository" "helm-test1" details page
```

sanketpathak commented on frontend/packages/helm-plugin/integration-tests/features/helm/helm-page-tabs.feature:
```diff
@@ -38,4 +38,54 @@ Feature: Add repositories tab in Helm navigation item
+               And user is at "Edit ProjectHelmChartRepository" form
+               And user enters Display name as "My charts"
+               And user clicks on Save button
+             Then user can see "ProjectHelmChartRepository" "helm-test1" details page
+               And user is at the Helm page
+               And user clicks on Repositories tab
+             Then user can see "ProjectHelmChartRepository" "helm-test1" updated with "My charts" in the list page
```
We can not have two Then statements

sanketpathak commented on frontend/packages/helm-plugin/integration-tests/support/step-definitions/helm/helm-navigation.ts:
```diff
@@ -6,16 +6,29 @@ import {
+  cy.byLegacyTestID('kebab-button').click();
+  cy.byTestActionID(`Delete ${type}`).click();
+  createForm.clickConfirm();
+  cy.wait(10000);
```
Can we avoid this using timeout

debsmita1 replied:
at times it takes some time for the resource to get deleted

jerolimov reviewed (approved):
Tested it locally and it works fine.

/approve
/lgtm

debsmita1 commented:
/label docs-approved
/label px-approved

invincibleJai commented:
/test e2e-gcp-console

sanketpathak commented:
Got this issue while testing on a cluster-bot
[Screencast from 08-06-2022 06:11:55 PM.webm](https://user-images.githubusercontent.com/10252230/183393162-883b3f37-4a9d-4869-967f-dfe93ba2d701.webm)

sanketpathak commented:
/label qe-approved
[Screencast from 08-09-2022 05:36:02 PM.webm](https://user-images.githubusercontent.com/10252230/183643710-4fc079ab-a54f-4371-a7c5-2022e3cd457b.webm)

sanketpathak commented:
/lgtm
